lib/srv/sess.go

```diff
@@ -1669,3 +1724,40 @@ func (s *session) trackSession(teleportUser string, policySet []*types.SessionTr
 
 	return nil
 }
+
+// emitAuditEvent emits audit events.
+func (s *session) emitAuditEvent(ctx context.Context, event apievents.AuditEvent) error {
+	rec := s.Recorder()
+	select {
+	case <-rec.Done():
+		newRecorder, err := newEventOnlyRecorder(s, s.scx)
+		if err != nil {
+			return trace.ConnectionProblem(err, "failed to recreate audit events recorder")
+		}
+		s.setRecorder(newRecorder)
+
+		return newRecorder.EmitAuditEvent(ctx, event)
+	default:
+		return rec.EmitAuditEvent(ctx, event)
+	}
+}
+
+// onWriteError defines the `OnWriteError` `TermManager` callback.
+func (s *session) onWriteError(idString string, err error) {
+	if idString == sessionRecorderID {
+		switch s.scx.Identity.RoleSet.SessionRecordingMode(constants.SessionRecordingServiceSSH) {
```

Comment:
The `TermManager` `[Write](https://github.com/gravitational/teleport/blob/f984976dfb1f2abecd514ee2d77d3431531606e6/lib/srv/termmanager.go#L103)/[BroadcastMessage](https://github.com/gravitational/teleport/blob/f984976dfb1f2abecd514ee2d77d3431531606e6/lib/srv/termmanager.go#L177)` functions use a lock to access the writer's list. Meaning that if we directly call a Writer inside this callback, it will be a deadlock; that's why both `BroadcastSystemMessage` and `Close` are called inside a goroutine.

After that, when the writer has an error, [it is removed from the `TermManager` writers list](https://github.com/gravitational/teleport/blob/f984976dfb1f2abecd514ee2d77d3431531606e6/lib/srv/termmanager.go#L92); in the case where we have only one `sessionRecorderID` this callback can only be called once.